pyadjoint/optimization/tao_solver.py

```diff
@@ -0,0 +1,491 @@
+from functools import cached_property
+from numbers import Complex
+import weakref
+
+import numpy as np
+
+from ..enlisting import Enlist
+from ..overloaded_type import OverloadedType
+from .optimization_problem import MinimizationProblem
+from .optimization_solver import OptimizationSolver
+
+
+try:
+    import petsc4py.PETSc as PETSc
+except ModuleNotFoundError:
+    PETSc = None
+
+__all__ = \
+    [
+        "TAOSolver"
+    ]
+
+
+def attach_destroy_finalizer(obj, *args):
+    def finalize_callback(*args):
+        for arg in args:
+            if arg is not None:
+                arg.destroy()
```

Comment:
@JDBetteridge says this needs to be `.delayed_destroy`.

Reply:
Not available, but the possible deadlock is now documented.

Reply:
This is intended to work around the following, which leaks memory on my machine (Ubuntu 24.04, run on two processes)
```
from firedrake import *
from firedrake.petsc import garbage_cleanup

import numpy as np

from itertools import count
import gc
import weakref


class Mat:
    def __init__(self, space):
        self._x = Function(space)
        self._test = TestFunction(space)
        self._memory = np.full(1000000, 1, dtype=np.uint8)

    def mult(self, x, y):
        with self._x.dat.vec_ro:
            x.copy(result=x_v)
        y_c = assemble(inner(self._x, self._test) * dx)
        with y_c.dat.vec_ro as y_v:
            y_v.copy(result=y)


def attach_destroy_finalizer(obj, *args):
    def finalize_callback(*args):
        for arg in args:
            if arg is not None:
                arg.destroy()

    finalize = weakref.finalize(obj, finalize_callback,
                                *args)
    finalize.atexit = False


mesh = UnitSquareMesh(10, 10)
space = FunctionSpace(mesh, "Lagrange", 1)
u = Function(space, name="u")
with u.dat.vec_ro as u_v:
    n, N = u_v.getSizes()


class Test:
    pass


for _ in count():
    t = Test()
    A = Mat(space)
    mat = PETSc.Mat().createPython(((n, N), (n, N)), A, comm=u.comm)
    # attach_destroy_finalizer(t, mat)  # Uncomment to fix leak

    del t
    gc.collect()
    garbage_cleanup(u.comm)
```
Arguably it would be better to track down the source of the leak -- but the case where I've hit this was much harder to debug than the above, hence the defensive use of `destroy`. The main downside is if properties are accessed and retained, e.g.
```
tao_solver = TAOSolver(...)
tao = tao_solver.tao
del tao_solver
# TaoDestroy has been called for tao
```

Reply:
The previous example leaks because `garbage_cleanup(u.comm)` doesn't call `PETSc.garbage_cleanup(u.comm)` (is this a separate bug?). However a bit more digging leads to this example, which leaks for me with two processes.
```
from firedrake import *
from firedrake.petsc import garbage_cleanup

import petsc4py.PETSc as PETSc

import gc
from itertools import count


class Mat:
    def mult(self, A, x, y):
        x.copy(result=y)


mesh = UnitSquareMesh(10, 10)
space = FunctionSpace(mesh, "Lagrange", 1)
u = Function(space, name="u")
with u.dat.vec_ro as u_v:
    n, N = u_v.getSizes()
b = Function(space).interpolate(Constant(1))


for i in count():
    A = Mat()
    mat = PETSc.Mat().createPython(((n, N), (n, N)), A, comm=u.comm)
    ksp = PETSc.KSP().createPython()
    ksp.setOperators(mat)
    ksp.setTolerances(rtol=0.1)
    ksp.setType(PETSc.KSP.Type.CG)
    with b.dat.vec as b_v, Function(space).dat.vec as u_v:
        ksp.solve(b_v, u_v)

    # Uncomment to fix leak
    # ksp.destroy()
    # mat.destroy()

    del mat, ksp

    # Increase to speed up leak
    j = 1
    if i % j == 0:
        print(f"{i=}", flush=True)
        gc.collect()
        garbage_cleanup(u.comm)
        PETSc.garbage_cleanup(u.comm)
```